matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/network/parsing/CredentialsResponseAdapter.kt

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2022 The Matrix.org Foundation C.I.C.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.internal.network.parsing
+
+import com.squareup.moshi.FromJson
+import org.matrix.android.sdk.api.auth.data.Credentials
+
+/**
+ * Used when we receive a new credential from the serer that contains an access token expiry. We want to
+ * estimate the expiry date as soon as possible using `expiresInMs`.
+ */
+class CredentialsResponseAdapter {
+    @FromJson
+    fun credentialsFromResponse(responseCredentials: Credentials): Credentials {
+        if (responseCredentials.expiresInMs == null || responseCredentials.expiryTs != null) {
+            // expiryTs already estimated or doesn't apply return existing credential
+            return responseCredentials
+        }
+        // We have received a credential response from the server estimate the expiry datetime.
+        return responseCredentials.copy(expiryTs = System.currentTimeMillis() + responseCredentials.expiresInMs)
```

Comment:
`System.currentTimeMillis()` is not ideal to manage datetime for this use case I think, since the phone date can be changed by the user. I have no better idea though, will think about it.

Reply:
Yea, the accuracy of the date isn't too important as we are just adding the expiry to whatever the user has set. So it will still expire at the time + `expiresInMs`. If the user does change the time forward in to the future and return to the app, it will likely just cause a refresh. Which is probably what happens most times you use the app anyway. So not sure there it's all that bad?

Reply:
The implementation is now complete and matches what we do on iOS. `expiresInMs` is used to preemptively estimate when the token needs to be updated, but ultimately we reply on the server returning `M_UNKNOWN_TOKEN` as the source of truth as to when we need to refresh. This translates to a [check before each request](https://github.com/vector-im/element-android/pull/5023/files#diff-47d9f959e736383577e4c16f04856294d5888805287086720b6d3dfbf7087261R37) to see if we need to preemptively refresh the token based on our estimate. But if the server returns a `M_UNKNOWN_TOKEN` for a request we also refresh the token and retry the original request [here](https://github.com/vector-im/element-android/pull/5023/files#diff-47d9f959e736383577e4c16f04856294d5888805287086720b6d3dfbf7087261R46).


So if the user moves:
 - moves the time into the future, it will cause a token refresh when we open the app, after which we will create a new estimate based on the response and the new time.
 - moves the time into the past, the expiry estimate will be incorrect, when the token expires the server will respond with an `M_UNKNOWN_TOKEN`, we will refresh and create a new estimate based on the response and the new time.

Let me know what you think.

Reply:
I think it is OK, thanks!